### ISSUE TITLE
query: PromQL resets function does not detect reset.

### ISSUE THREAD

**Issue by bwplotka** [thanos-io/thanos#3644]
Just for awareness. I noticed PromQL `resets` https://prometheus.io/docs/prometheus/latest/querying/functions/#resets function has no right to work. (: 

This bug was introduced  ~18.05.2020 by http://github.com/thanos-io/thanos/pull/2548

This is because for deduplication and downsampling precision we adjust resets on storage level: 
* Query:  https://github.com/thanos-io/thanos/blob/49dad93ed67210374939aedd7a8f849a60375906/pkg/query/iter.go#L188
* Downsampling: https://github.com/thanos-io/thanos/blob/47f9a225cc525bb0c7669c77a7a16a52f0333cfc/pkg/compact/downsample/downsample.go#L528

Since it's used in downsampling it irreversibly blocked the reset function against such downsampled data.
 
First of all, I wonder... Why someone would use such low level detail like when counter resets and how many were happening? Maybe to detect restarts? 🤔 

Still, I think we should fix it, help wanted 🤗 I noticed this when adding those tests https://github.com/thanos-io/thanos/pull/3631

**Comment by bwplotka** [comment]
Some useful disscussion on IRC:

```
bbrazil
resets is intended largely for debugging broken counters, however some people have found uses for it with gauges too
bwplotka
per: https://github.com/thanos-io/thanos/issues/3644
bbrazil
as resets will catch any time a series decreased, so can catch the down of a flap
bwplotka
yea ok
thanks for info (:
bbrazil
I'm having difficultly imaginging a good use case for using it on downsampled data, but that doesn't mean one doesn't exist
probably use the average or last rather than the counter. and it should be okay
bwplotka
what do you mean? For "resets" results? Interesting
bbrazil
basically resets is a gauge function, not a counter function
```

**Comment by bwplotka** [comment]
```
# Testdata for resets() and changes().
load 5m
	http_requests{path="/foo"}	1 2 3 0 1 0 0 1 2 0
	http_requests{path="/bar"}	1 2 3 4 5 1 2 3 4 5
	http_requests{path="/biz"}	0 0 0 0 0 1 1 1 1 1

# Tests for resets().
eval instant at 50m resets(http_requests[5m])
	{path="/foo"} 0
	{path="/bar"} 0
	{path="/biz"} 0

eval instant at 50m resets(http_requests[20m])
	{path="/foo"} 1
	{path="/bar"} 0
	{path="/biz"} 0

eval instant at 50m resets(http_requests[30m])
	{path="/foo"} 2
	{path="/bar"} 1
	{path="/biz"} 0

eval instant at 50m resets(http_requests[50m])
	{path="/foo"} 3
	{path="/bar"} 1
	{path="/biz"} 0

eval instant at 50m resets(nonexistent_metric[50m])
```
Tests that detect those.

**Comment by ee-github** [comment]
I would like to make a case here for the usefulness of fixing this bug, further and similar to the use-cases above mentioned by Brian Brazil.

When doing blackbox style monitoring of monotonically increasing functions, resets() is very valuable in detecting issues.

For example, uptime counter of network routing protocol like BGP.  It ticks up until the session resets, at which point it restarts from zero.  The resets() function catches that very well.

Just monitoring the binary/bool "up" status of said BGP peer is not sufficient because it can bounce quickly in-between scrape intervals and go undetected.  However, the such quick events cannot escape resets() of an uptime counter.  Very valuable.

Please do fix!  Beer 🍺 on me for the contributor!
